src/vivarium/framework/values.py

```diff
@@ -193,15 +229,66 @@ class Pipeline:
 
     """
 
-    def __init__(self):
-        self.name = None
-        self.source = None
-        self.mutators = []
-        self.combiner = None
-        self.post_processor = None
-        self.manager = None
+    def __init__(self) -> None:
+        self.name: str | None = None
+        self._source: ValueSource | None = None
+        self.mutators: list[ValueMutator] = []
+        self._combiner: ValueCombiner | None = None
+        self.post_processor: PostProcessor | None = None
+        self._manager: ValuesManager | None = None
+
+    def _get_attr_error(self, attribute: str) -> str:
+        return (
+            f"The pipeline for {self.name} has no {attribute}. This likely means "
+            f"you are attempting to modify a value that hasn't been created."
+        )
+
+    def _set_attr_error(self, attribute: str, new_value: Any) -> str:
+        current_value = getattr(self, f"_{attribute}")
+        return (
+            f"A second component is attempting to set the {attribute} for pipeline {self.name} "
+            f"with {new_value}, but it already has a {attribute}: {current_value}."
+        )
+
+    @property
+    def source(self) -> ValueSource:
+        if self._source is None:
+            raise DynamicValueError(self._get_attr_error("source"))
+        return self._source
+
+    @source.setter
+    def source(self, source: ValueSource) -> None:
+        if self._source is not None:
+            raise DynamicValueError(self._set_attr_error("source", source))
+        self._source = source
+
+    @property
+    def combiner(self) -> ValueCombiner:
+        if self._combiner is None:
+            raise DynamicValueError(self._get_attr_error("combiner"))
+        return self._combiner
+
+    @combiner.setter
+    def combiner(self, combiner: ValueCombiner) -> None:
+        if self._combiner is not None:
+            raise DynamicValueError(self._set_attr_error("combiner", combiner))
+        self._combiner = combiner
+
+    @property
+    def manager(self) -> "ValuesManager":
+        if self._manager is None:
+            raise DynamicValueError(self._get_attr_error("manager"))
+        return self._manager
+
+    @manager.setter
+    def manager(self, manager: "ValuesManager") -> None:
+        if self._manager is not None:
+            raise DynamicValueError(self._set_attr_error("manager", manager))
+        self._manager = manager
```

Comment:
Woah this is working because of the different decorators? Didn't know you could do that